Reader Comments: Implement comment rendering via WPRichContent

dvdchr opened the pull request:
Refs #17476
Depends on #17634

As titled, this PR reintroduces the previous content rendering method through `WPRichContentView` into the comment thread.  Note that all pre-existing issues with the `WPRichContentView` (e.g. enlarged emojis, extra white spaces, partially clipped text) are not fixed here.

Here's a preview:
![Simulator Screen Shot - iPhone 12 - 2021-12-07 at 22 09 36](https://user-images.githubusercontent.com/1299411/145059161-1df55af6-f6f8-4846-9581-ff1db81fda9e.png)

---

I also fixed a minor visual issue regarding table view separators. Previously the separators were hidden – but now since we are showing a separator, the separators between empty cells show while the loading view is shown. This PR fixes that issue.

Before | After
--|--
![before](https://user-images.githubusercontent.com/1299411/145061521-83927ac7-5611-4dc8-a6b4-70e22a258c25.png) | ![after](https://user-images.githubusercontent.com/1299411/145061534-7d853db8-a485-4cdc-b7af-7b8d76044908.png)


## To test

- Ensure that the `newCommentThread` flag is enabled.
- Go to Reader > any comment thread.
- 🔍 Verify that the comment body is displayed with the old appearance, but contained in the new comment cell.
- 🔍 Tap on the overflow menu and verify that the menu items work.
- 🔍 Verify that the reply button works.
- 🔍 Verify that the like button works.
- Find a comment that contains a link in it.
- 🔍  Tap the link and verify that a modal window is presented and it loads the tapped URL.
- Find a comment that contains an image (or some).
- 🔍  Tap on the image and verify that a modal window is presented and it loads the image.

#### Verifying the table separator issue

- Go to Notifications, and tap any comment reply notification.
- Tap the notification header.
- 🔍 Verify that no table separators are shown while the loading view is visible. 

## Regression Notes
1. Potential unintended areas of impact
n/a. Feature is hidden behind a flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a. Feature is hidden behind a flag.

3. What automated tests I added (or what prevented me from doing so)
n/a. Feature is hidden behind a flag.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Implement comment rendering with WPRichContentView

dvdchr committed:
Add an important warning on estimatedHeight method

dvdchr committed:
Update docs

dvdchr committed:
Hide table separators when loading view is shown

dvdchr committed:
Simplify fix for the table separator

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -170,7 +179,7 @@ class CommentContentTableViewCell: UITableViewCell, NibReusable {
     ///   - comment: The `Comment` object to display.
     ///   - onContentLoaded: Callback to be called once the content has been loaded. Provides the new content height as parameter.
-    func configure(with comment: Comment, onContentLoaded: ((CGFloat) -> Void)?) {
+    func configure(with comment: Comment, renderMethod: RenderMethod = .web, onContentLoaded: ((CGFloat) -> Void)?) {
```
It'd be nice to add `renderMethod` to the params list in the method description above.

dvdchr replied:
Whoops! I missed this one. It's now addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17635/commits/54d99efd3249066902638b9bc55b9df240a85f94. Thanks!

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -214,13 +223,24 @@ class CommentContentTableViewCell: UITableViewCell, NibReusable {
     }
 
+    @objc func ensureRichContentTextViewLayout() {
+        guard let renderer = renderer,
```
This throws a warning:
`Immutable value 'renderer' was never used; consider replacing with '_' or removing it`

dvdchr replied:
Thank you! I forgot about this. This is now fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17635/commits/54d99efd3249066902638b9bc55b9df240a85f94.

ScoutHarris commented:
It looks like the layout isn't sized quite right sometimes.

Moderation in post threaded comments:
- In the threaded view, moderate a comment. It disappears as expected.
- Select the My Site tab (i.e. don't navigate back via the nav bar). 
- Go to Comments and approve the comment.
- Select the Reader tab. 
- The comment does reappear in the threaded view, but sometimes the UI isn't quite right. I couldn't repro it 100% of the time, but I did see it a couple of times.

<img width="777" alt="Screen Shot 2021-12-07 at 1 11 37 PM" src="https://user-images.githubusercontent.com/1816888/145103526-5c20e547-1f3a-4bde-8735-09b7d99e4cfd.png">


Notification Comments:
- On a mention comment notification, tap the header to show the threaded comments.
- Sometimes the reaction bar is cut off or duplicated:


<img width="471" alt="Screen Shot 2021-12-07 at 1 50 15 PM" src="https://user-images.githubusercontent.com/1816888/145104545-db03f79c-9d12-481e-92a4-b555c60bc299.png">

It seems to only happen when the comments are initially loaded because after the first viewing the UI is fine. But it is 100% reproducible when the comments are reloaded.

ScoutHarris commented:
I can repro this with the old threaded comments, so it pre-exists your changes. But it would be nice to fix this at some point. Sometimes the loading view still appears after the comments are loaded:

![Simulator Screen Shot - iPhone 12 Pro Max - 2021-12-07 at 13 32 29](https://user-images.githubusercontent.com/1816888/145105089-b5870e5d-f2cc-4e0b-91e8-1998f8070794.png)

![Simulator Screen Shot - iPad Pro (9 7-inch) - 2021-12-07 at 13 16 54](https://user-images.githubusercontent.com/1816888/145105143-8d885ca0-f00d-4195-b38d-478e4bd04de0.png)

dvdchr committed:
Update docs and fix warning

dvdchr committed:
Increase horizontal hugging priority for the name label

In some cases, the author label could appear "pushed" to the edge of the container, creating space between the badge and the name label.

dvdchr committed:
Update NRV placement in Reader Comments for iPad

dvdchr committed:
Merge remote-tracking branch 'origin/develop' into issue/17476-rich-content-rendering

dvdchr committed:
Fix cell overlap issue

dvdchr commented:
Thank you for the details! 👍 

> Sometimes the loading view still appears after the comments are loaded:

This should be fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17635/commits/a7b7fcd8c3fdb1d5f15050b4a65eaf163fa5f5b9. With this, I've also addressed the wrong NRV placement in iPad. Here's a preview:

• | Before | After
--|--|--
Portrait | ![ipad_portrait_before](https://user-images.githubusercontent.com/1299411/145353279-bdc2802b-2e1e-48ad-a78d-506d2093f5ea.png) | ![ipad_portrait_after](https://user-images.githubusercontent.com/1299411/145353287-27bd7e37-28d4-4a83-9604-6f38f32f0094.png)
Landscape | ![ipad_landscape_before](https://user-images.githubusercontent.com/1299411/145353268-bd04e603-e480-4042-9764-4027f595b8f6.png) | ![ipad_landscape_after](https://user-images.githubusercontent.com/1299411/145353283-5df889bd-f611-4cee-8a3c-45ba0ea35bbe.png)

> Sometimes the reaction bar is cut off or duplicated:

Yeah, it looks like this is a bug that has been haunting our threads 😅 . I've investigated this and filed an issue here: https://github.com/wordpress-mobile/WordPress-iOS/issues/17653. I've also attempted to address this in https://github.com/wordpress-mobile/WordPress-iOS/pull/17635/commits/7d2fad2ffbc3ac84bce8b7b1b22ae30ec6b1c615, since we're going to turn this on for the public anyways this upcoming release and I'd want to prevent side effects when we have to turn off the flag.

Here's a preview:

• | Before | After
--|--|--
Portrait | ![17635_ipad_portrait_before](https://user-images.githubusercontent.com/1299411/145352518-54841e39-ae4d-4756-baa5-3a9c4235ff84.png) | ![17635_ipad_portrait_after](https://user-images.githubusercontent.com/1299411/145352496-8f515187-9032-4351-a440-e93dae267452.png)
Landscape | ![17635_ipad_landscape_before](https://user-images.githubusercontent.com/1299411/145352515-58bfd22a-7060-4849-b3eb-9ca94711ab78.png) |  ![17635_ipad_landscape_after](https://user-images.githubusercontent.com/1299411/145352867-4d30684b-4c1f-479c-8813-89ef0d74fc40.png)

As for the case with moderation and the Reader tab, I've been unable to reproduce it so far.  If this still appears, I'll give it a go again next week. Ready for your next round of review, please! 🙂

dvdchr commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:
```diff
@@ -178,7 +178,6 @@ private extension WPRichContentView {
             "em, i { font:-apple-system-body; font-family: 'Noto Serif'; font-weight: normal; font-style: italic; line-height:1.6; } " +
             "a { color: #\(style.linkColorHex); text-decoration: none; } " +
             "a:active { color: #\(style.linkColorActiveHex); } " +
-            "img.emoji { vertical-align: baseline; max-width: 1rem; } " +
```
This was an experiment that I forgot to clean up. This line doesn't do anything to the `WPRichContentView`. 😅

ScoutHarris commented:
Well, I see the NRV is misplaced on iPad:

<kbd>![ipad](https://user-images.githubusercontent.com/1816888/145453615-cabe9ba4-8706-4d09-aef3-c44ed7e74e05.png)</kbd>

It could be fixed separately if you like since it's not really related to this PR.

dvdchr committed:
Fix NRV misalignment on iPad

dvdchr commented:
> Well, I see the NRV is misplaced on iPad:

Thank you! I was only testing this on iOS 14.5 and couldn't figure out why it isn't happening on my end. Turns out, it happens on iOS 15 – and guess what: iOS 15 decides to call `traitCollectionDidChange:` on push navigation. The `ReaderCommentsViewController` happens to call `[self refreshTableViewAndNoResultsView]` on this method, which is the root cause of the misalignment. Life of an iOS engineer, indeed. 😄 

It should be fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17635/commits/eea5e108442aac01e425992a7ef3db35ddaed21f, and I'll merge this ahead for now. If there are any further issues, I'll address them through a separate PR. 🙂
